rosbag2_cpp/src/rosbag2_cpp/cache/cache_consumer.cpp

```diff
@@ -70,12 +69,12 @@ void CacheConsumer::exec_consuming()
     // swap producer buffer with consumer buffer
     message_cache_->wait_for_buffer();
 
-    // make sure to use consistent callback for each iteration
-    auto callback_for_this_loop = consume_callback_;
-
     // consume all the data from consumer buffer
```

Comment:
the point for the callback_for_this_loop was performance. We changed the current callback in crucial producer (subscription processing) thread and moved on, instead of waiting for the whole buffer flush to finish.

Reply:
I see what you are saying.  In the case where this thread is currently writing I/O, this could block the `change_consume_callback` for a while until that is finished.

What do you think if I changed this to:

```
consume_callback_function_t callback_for_this_loop;
{
std::lock_guard<std::mutex> lg(consumer_mutex_);
callback_for_this_loop = consume_callback_
}
auto consumer_buffer = message_cache_->consumer_buffer();
callback_for_this_loop(consumer_buffer->data());
```

That way we only hold up `change_consume_callback` for a short while (just enough to copy the `std::function`), but we still retain the property that we don't need to call `close`.

Reply:
This looks good! Now I am trying to remember why this section:

```
if (!consumer_thread_.joinable()) {
    is_stop_issued_ = false;
    consumer_thread_ = std::thread(&CacheConsumer::exec_consuming, this);
  }
```
was there. Plus I will stress-test the change with our benchmarks, which I can do tomorrow, to see if all is ok.

Reply:
OK, great.  I'll make that change and wait to hear back from you how it does in the benchmarks.  Thanks for your support.

Reply:
All right, changed as discussed in 6e137ce .

Reply:
I explain why this section is necessary and also ran benchmarks of this PR:
1) without the section (PR as is) - failure to record anything post 1st bag
2) with the section restored - all correct and looking good